comp/rdnsquerier/impl/rdnsquerier.go

```diff
@@ -167,18 +169,107 @@ func (q *rdnsQuerierImpl) GetHostname(ipAddr []byte, updateHostnameSync func(str
 	if err != nil {
 		q.logger.Debugf("Reverse DNS Enrichment cache.getHostname() for addr %s returned error: %v", netipAddr.String(), err)
-		return err
 	}
 
-	return nil
+	return err
+}
+
+// GetHostname attempts to resolve the hostname for the given IP address synchronously.
+// If the IP address is invalid then an error is returned.
+// If the IP address is not in the private address space then it is ignored - no lookup is performed and nil error is returned.
+// If the IP address is in the private address space then the IP address will be resolved to a hostname.
+// The function accepts a timeout via context and will return an error if the timeout is reached.
+func (q *rdnsQuerierImpl) GetHostname(ctx context.Context, ipAddr string) (string, error) {
+	results := q.GetHostnames(ctx, []string{ipAddr})
+	result, ok := results[ipAddr]
+	if !ok {
+		return "", fmt.Errorf("no result for IP address %s", ipAddr)
+	}
+
+	return result.Hostname, result.Err
+}
+
+// GetHostnames attempts to resolve the hostname for the given IP addresses.
+// If the IP address is invalid then an error is returned.
+// If the IP address is not in the private address space then it is ignored - no lookup is performed and nil error is returned.
+// If the IP address is in the private address space then the IP address will be resolved to a hostname.
+// The function accepts a timeout via context and will return an error if the timeout is reached.
+func (q *rdnsQuerierImpl) GetHostnames(ctx context.Context, ipAddrs []string) map[string]rdnsquerier.ReverseDNSResult {
+	q.internalTelemetry.total.Add(float64(len(ipAddrs)))
+
+	var wg sync.WaitGroup
+	resultsChan := make(chan rdnsquerier.ReverseDNSResult, len(ipAddrs))
+
+	for _, ipAddr := range ipAddrs {
+		wg.Add(1)
+		go func(ipAddr string) {
+			defer wg.Done()
+			netipAddr, err := netip.ParseAddr(ipAddr)
+			if err != nil {
+				q.internalTelemetry.invalidIPAddress.Inc()
+				resultsChan <- rdnsquerier.ReverseDNSResult{IP: ipAddr, Err: fmt.Errorf("invalid IP address %s: %v", ipAddr, err)}
+				return
+			}
+
+			if !netipAddr.IsPrivate() {
+				q.logger.Tracef("Reverse DNS Enrichment IP address %s is not in the private address space", ipAddr)
+				resultsChan <- rdnsquerier.ReverseDNSResult{IP: ipAddr}
+				return
+			}
+			q.internalTelemetry.private.Inc()
+
+			hostnameChan := make(chan string, 1)
+			asyncErrChan := make(chan error, 1)
+
+			err = q.cache.getHostname(
+				netipAddr.String(),
+				func(h string) {
+					hostnameChan <- h
+					asyncErrChan <- nil
+				},
+				func(h string, e error) {
+					hostnameChan <- h
+					asyncErrChan <- e
+				},
+			)
+			if err != nil {
+				q.logger.Debugf("Reverse DNS Enrichment cache.getHostname() for addr %s returned error: %v", netipAddr.String(), err)
+			}
+
+			select {
+			case hostname := <-hostnameChan:
+				asyncErr := <-asyncErrChan // this is okay because we know that as soon as we send hostname, we send asyncErr
```

Comment:
Not sure this is completely ok, because the async callback provided to q.cache.getHostname() can be processed asynchronously by multiple workers I think there's the possibility of a race condition when resolving multiple IP addresses:

1. IP1 is resolved without error, async callback for IP1 does hostnameChan <- h
2. IP2 is not resolved, with error, async callback for IP2 does hostnameChan <- h
3. async callback for IP2 does asyncErrChan <- e
4. async callback for IP1 does asyncErrChan <- e
and at this point the hostname read from hostnameChan is the hostname of IP1 and the asyncError read from asyncErrChan is the error encountered resolving IP2.  Why not use a single channel to return both hostname/error?

Reply:
Nevermind, I wasn't paying attention, should be fine because unique hostnameChan and asyncErrorChan used for each ip addr.